File: sdk-api/src/test/java/dev/restate/sdk/RestateCodegenTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import dev.restate.sdk.core.RestateCodegenTestSuite;
import dev.restate.sdk.core.testservices.*;
import io.grpc.BindableService;
import java.time.Duration;

public class RestateCodegenTest extends RestateCodegenTestSuite {

  private static class GreeterWithRestateClientAndServerCodegen
      extends GreeterRestate.GreeterRestateImplBase {

    @Override
    public GreetingResponse greet(RestateContext context, GreetingRequest request) {
      GreeterRestate.GreeterRestateClient client = GreeterRestate.newClient(context);
      client.delayed(Duration.ofSeconds(1)).greet(request);
      client.oneWay().greet(request);
      return client.greet(request).await();
    }
  }

  @Override
  protected BindableService greeterWithRestateClientAndServerCodegen() {
    return new GreeterWithRestateClientAndServerCodegen();
  }

  private static class Codegen extends CodegenRestate.CodegenRestateImplBase {

    @Override
    public MyMessage emptyInput(RestateContext context) {
      CodegenRestate.CodegenRestateClient client = CodegenRestate.newClient(context);
      return client.emptyInput().await();
    }

    @Override
    public void emptyOutput(RestateContext context, MyMessage request) {
      CodegenRestate.CodegenRestateClient client = CodegenRestate.newClient(context);
      client.emptyOutput(request).await();
    }

    @Override
    public void emptyInputOutput(RestateContext context) {
      CodegenRestate.CodegenRestateClient client = CodegenRestate.newClient(context);
      client.emptyInputOutput().await();
    }

    @Override
    public MyMessage oneWay(RestateContext context, MyMessage request) {
      CodegenRestate.CodegenRestateClient client = CodegenRestate.newClient(context);
      return client._oneWay(request).await();
    }

    @Override
    public MyMessage delayed(RestateContext context, MyMessage request) {
      CodegenRestate.CodegenRestateClient client = CodegenRestate.newClient(context);
      return client._delayed(request).await();
    }
  }

  @Override
  protected BindableService codegen() {
    return new Codegen();
  }
}

File: sdk-api-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/RestateCodegenTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.RestateCodegenTestSuite
import dev.restate.sdk.core.testservices.*
import dev.restate.sdk.core.testservices.CodegenRestateKt.CodegenRestateKtImplBase
import dev.restate.sdk.core.testservices.GreeterRestateKt.GreeterRestateKtImplBase
import io.grpc.BindableService
import kotlin.time.Duration.Companion.seconds

class RestateCodegenTest : RestateCodegenTestSuite() {
  private class GreeterWithRestateClientAndServerCodegen : GreeterRestateKtImplBase() {
    override suspend fun greet(context: KeyedContext, request: GreetingRequest): GreetingResponse {
      val client = GreeterRestateKt.newClient(context)
      client.delayed(1.seconds).greet(request)
      client.oneWay().greet(request)
      return client.greet(request).await()
    }
  }

  override fun greeterWithRestateClientAndServerCodegen(): BindableService {
    return GreeterWithRestateClientAndServerCodegen()
  }

  private class Codegen : CodegenRestateKtImplBase() {
    override suspend fun emptyInput(context: Context): MyMessage {
      val client = CodegenRestateKt.newClient(context)
      return client.emptyInput().await()
    }

    override suspend fun emptyOutput(context: Context, request: MyMessage) {
      val client = CodegenRestateKt.newClient(context)
      client.emptyOutput(request).await()
    }

    override suspend fun emptyInputOutput(context: Context) {
      val client = CodegenRestateKt.newClient(context)
      client.emptyInputOutput().await()
    }

    override suspend fun oneWay(context: Context, request: MyMessage): MyMessage {
      val client = CodegenRestateKt.newClient(context)
      return client._oneWay(request).await()
    }

    override suspend fun delayed(context: Context, request: MyMessage): MyMessage {
      val client = CodegenRestateKt.newClient(context)
      return client._delayed(request).await()
    }
  }

  override fun codegen(): BindableService {
    return Codegen()
  }
}
